py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,502 @@
+                 'start_date': event_dict['start_date'],
+                 'end_date': event_dict['end_date'],
+                 'format_timer': time_to_composite})
+
+            self.py3.composite_update(event_formatted, {'index': index})
+            responses.append(event_formatted)
+
+            index += 1
+
+            self.no_update = False
+
+        format_separator = self.py3.safe_format(self.format_separator)
+        self.py3.composite_update(format_separator, {'index': 'sep'})
+        responses = self.py3.composite_join(format_separator, responses)
+
+        return {'events': responses}
+
+    def google_calendar(self):
+        """
+        The method that outputs the response.
+
+        First, we check credential authorization. If no authorization, we display an error message,
+        and try authorizing again in 5 seconds.
+
+        Otherwise, we fetch the events, build the response, and output the resulting composite.
+        """
+        composite = {}
+
+        if not self.is_authorized:
+            cached_until = 0
+            self.is_authorized = self._authorize_credentials()
+        else:
+            if not self.no_update:
+                self.events = self._get_events()
+
+            composite = self._build_response()
+            cached_until = self.cache_timeout
+
+        return {
+            'cached_until': self.py3.time_in(cached_until),
+            'composite': self.py3.safe_format(self.format, composite)
+        }
+
+    def on_click(self, event):
+        if self.is_authorized and self.events is not None:
+            """
+            If button_refresh is clicked, we allow the events to be updated if the
+            last event update occured at least 1 second ago. This prevents a bug that
+            can crash py3status since refreshing the module too fast results in
+            incomplete event information being fetched as _get_events() is called
+            repeatedly.
+
+            Otherwise, we disable event updates.
+            """
+            try:
+                self.no_update = True
+                button = event['button']
+                button_index = event['index']
+
+                if button_index == 'sep':
+                    self.py3.prevent_refresh()
+                elif button == self.button_refresh:
+                    now = datetime.datetime.now()
+                    diff = (now - self.last_update).seconds
+                    if diff > 1:
+                        self.no_update = False
+                elif button == self.button_toggle:
+                    self.button_states[button_index] = \
+                        not self.button_states[button_index]
+                elif button == self.button_open:
+                    self.py3.command_run('xdg-open ' + self.event_urls[button_index])
+                    self.py3.prevent_refresh()
+                else:
+                    self.py3.prevent_refresh()
+            except IndexError:
+                return
```

Comment:
Do we still need `IndexError` here? What's tripping that? Also, we ought to analyze / understand `self.no_update` issue better too. Do we still need `self.no_update`? Because I think we didn't have any `self.py3.prevent_refresh()` in place before.

Reply:
I forgot I was using `IndexError` to try to handle earlier issues with clicking on separators. Shouldn't be needed anymore. As for `self.no_update` it's still an issue.

When we click a button, it's set to `True`. We stop the module from fetching events using the google calendar API. We don't set it to `False` until we have built a response with `_build_response()` from the events we already have. Why? If we fetch events from the calendar too fast, the module will crash. 

Imagine `self.no_update` isn't there. If we don't restrict how often we fetch events from the calendar, then the user can press `button_refresh` many times in a short period of time, which will cause events to be repeatedly fetched, eventually crashing the module (and probably py3status, too).